### PR TITLE
Avoid copies in matmul

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76828

With this PR, matmul just folds a bmm into a mm o mv if and only if it
can achieve so without copying. We add tests for this to make sure that
our algorithm to detect this is accurate.

For the cases where it was copying before see https://github.com/pytorch/pytorch/pull/75197#discussion_r843413208 https://github.com/pytorch/pytorch/pull/75197#discussion_r863489479 https://github.com/pytorch/pytorch/pull/75197#discussion_r863489805

Fixes https://github.com/pytorch/pytorch/issues/76702

cc @ngimel @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

**Comment by mruberry** [comment]
Offline feedback:

- which implementation is more accurate, and why are tests failing?

**Comment by Lezcano** [comment]
Summary of the state of affairs:
We have four CI jobs failing. [CUDA 10.2 py 3.9 linux](https://github.com/pytorch/pytorch/runs/7321226034?check_suite_focus=true), [CUDA 11.6 py 3.7 linux](https://github.com/pytorch/pytorch/runs/7320800945?check_suite_focus=true), [CUDA 11.6 win](https://github.com/pytorch/pytorch/runs/7321114098?check_suite_focus=true) and [CUDA 11.6 py 3.7 linux (other)](https://github.com/pytorch/pytorch/runs/7320089217?check_suite_focus=true).
They all fail in the same test. A test for the multihead attention implementation in core, namely `test_native_multihead_self_attention_cuda_float*`. There is a small reasonable failure in `float32`, and a larger unreasionable failure in `float16`.

About these tests, it's worth noting that they compare two different implementations of the same function, namely `torch.nn.MultiheadAttention` and `torch._native_multi_head_attention`. They compare them by feeding them the same inputs on the same dtype, and checking that they are close enough. Thus, a failure on the test for `float16` may mean that this PR worsens OR improves the result of `matmul` for `float16`---it's not clear a priori which. For context, the matrices being used in the relevant matmul are of shapes `[16, 8, 64] x [64, 64]`.

Let's start with the easier one. The `float32` failure I'm able to reproduce locally, but it's a difference of 2.4e-5 on one of the elements. IMO, this is an acceptable difference. Also, when modifying a bit the test, the error stops being reproducible, so yeah.

The interesting part is the `float16` test. This one I am not able to reproduce locally (the tests simply pass). This is in an RTX 2060 with CUDA 11.4.

I do not know how to follow up on this, as this seems to be coming from a subtle bug from cublas/cusolver in some versions. CC @xwang233 @IvanYashchuk for visibility, and @ngimel for advice on how to proceed.

**Comment by ngimel** [comment]
I agree that fp32 failure is reasonable. THe difference in fp16 is likely due to different hardware - CI is run with old M60s.
We could rewrite test to compare that the results are not further from the ground truth reference computation using fp32, instead of comparing fp16 results, maybe that'll satisfy CI. It's encouraging that on newer hardware it passes.

**Comment by xwang233** [comment]
@Lezcano what's the action needed on cublas and cusolver side? Do you have a C++ repro that demonstrates the failure in the libraries?

**Comment by Lezcano** [comment]
@xwang233 I'll try to put together a small repro so that the nvidia team can have a look at it

**Comment by Lezcano** [comment]
@xwang233 @IvanYashchuk I have tried for some time to get a small repro, but pulling out the inner computations in the test that happen within `matmul` is proving a bit tricky, given that they don't repro in my local machine and I need to ssh into the job, which is a pain.

The job is running on Tesla M60, and the nvidia version seems to be irrelevant, as the failures appear in 10.2 and 11.6. Could any of you try to repro these in a GPU with similar capabilities if you have access to any? The failing tests are `test/test_native_mha.py`, and the code to pull out the tensors within ATen is the following:
```c++
#include <torch/torch.h>
#include <iostream>
#include <vector>

int main() {
    auto t = torch::ones({3, 3});
    auto data = torch::pickle_save(t);
    std::ofstream file("tensor.pt");
    file.write(data.data(), data.size());
    std::cout << "Done!\n";
}
```
and then yo may load it as
```python
import torch

with open('tensor.pt', 'rb') as f:
    t = torch.load(f)
    print(t)
```

The difference comes from going through `mm` or `bmm` for tensors that appear within this test. 
As such, I was going to serialise `tensor1` and `tensor2` within `matmul` and test the differences in the results from going through the `should_fold` branch, to going through the `else` branch. One of them is clearly off from the result if you run `bmm` (or folded `mm`) in `float64`. Then this error may be tracked down to the current implementation within cublas.

Does this make sense? Please ping me if I can help any further.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebra.cpp:1742]:
```suggestion
      return true;
```

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebra.cpp:1815]:
```suggestion
    const auto t1_folded = t1->reshape({folded_dim1, sizes_1.back()});
```

**Review by IvanYashchuk — COMMENTED** [review]
Well, I was able to reproduce the test failures on M60 GPU.
Changing two lines of code makes the tests pass. But I don't have ideas how to really resolve the problem.
Also, I noticed in the logs test failures not with the element mismatch but with a runtime error:
```py
RuntimeError: CUDA driver API confirmed a leak in __main__.TestMHADeviceTypeCUDA.test_native_multihead_self_attention_cuda_float32! Caching allocator allocated memory was 0 and is now reported as 370688 on device 0. CUDA driver allocated memory was 101449728 and is now 103546880.
```

**Comment by Lezcano** [comment]
Thank you for the analysis Ivan! And yep, those changes would pretty much invalidate the whole point of this PR.

@ngimel for reference, @IvanYashchuk tried it in a Titan 1080 and the tests were passing, so this is not something that affects even all the maxwell cards, but, as far as we know, it may just affect M60s.

In my opinion, given that this is such an old and niche card (Ivan even had issues finding access to one of them within Nvidia), we should merge the PR and xfail these tests. Now, this would mean that we would need to xfail pretty much all the multihead attention tests on float16 and bfloat16. Other possibility would be to execute these tests in other CI box with a different GPU. Would that be possible?

**Comment by janeyx99** [comment]
Looping in @seemethere--how does the process of upgrading the GPUs in our CI runners go?

**Comment by IvanYashchuk** [comment]
I'm sorry to report that `test_native_mha.py` also fails on V100, P100, and works on A100:

<details>
  <summary>Test failures log</summary>

  
  ```py
======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=False, pad_all=False, use_nt=False, need_weights=False, average_attn_weights=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 51 / 8192 (0.6%)
Greatest absolute difference: 0.02392578125 at index (2, 3, 52) (up to 0.001 allowed)
Greatest relative difference: 0.032996632996632996 at index (2, 3, 52) (up to 0.001 allowed)

======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=False, pad_all=False, use_nt=True, need_weights=False, average_attn_weights=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 4 / 8192 (0.0%)
Greatest absolute difference: 0.00244140625 at index (6, 7, 35) (up to 0.001 allowed)
Greatest relative difference: 0.06442307692307692 at index (6, 7, 19) (up to 0.001 allowed)

======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=False, pad_all=False, use_nt=True, need_weights=False, average_attn_weights=True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 46 / 8192 (0.6%)
Greatest absolute difference: 0.023193359375 at index (12, 3, 3) (up to 0.001 allowed)
Greatest relative difference: 0.2107750472589792 at index (12, 3, 38) (up to 0.001 allowed)

======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=False, pad_all=False, use_nt=True, need_weights=True, average_attn_weights=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 28 / 8192 (0.3%)
Greatest absolute difference: 0.009765625 at index (8, 6, 51) (up to 0.001 allowed)
Greatest relative difference: 0.6784530386740332 at index (8, 6, 37) (up to 0.001 allowed)

======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=True, pad_all=False, use_nt=False, need_weights=True, average_attn_weights=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 255, in _test_multihead_attention_impl
    torch.testing.assert_close(weight_pt, weight_npt)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 4096 (0.0%)
Greatest absolute difference: 6.29425048828125e-05 at index (6, 2, 7, 2) (up to 1e-05 allowed)
Greatest relative difference: 0.03166986564299424 at index (6, 2, 7, 2) (up to 0.001 allowed)

======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=True, pad_all=False, use_nt=True, need_weights=False, average_attn_weights=True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 8192 (0.0%)
Greatest absolute difference: 0.001953125 at index (6, 2, 46) (up to 0.001 allowed)
Greatest relative difference: 0.0024922118380062306 at index (6, 2, 46) (up to 0.001 allowed)

======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=True, pad_all=False, use_nt=True, need_weights=True, average_attn_weights=True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 18 / 8192 (0.2%)
Greatest absolute difference: 0.005859375 at index (12, 1, 18) (up to 0.001 allowed)
Greatest relative difference: 0.10690121786197564 at index (12, 1, 43) (up to 0.001 allowed)

======================================================================
FAIL: test_native_multihead_self_attention_cuda_float16 (__main__.TestMHADeviceTypeCUDA) (use_padding=True, pad_all=True, use_nt=False, need_weights=False, average_attn_weights=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 3 / 8192 (0.0%)
Greatest absolute difference: 0.0020751953125 at index (1, 5, 4) (up to 0.001 allowed)
Greatest relative difference: 0.008803728638011392 at index (1, 5, 4) (up to 0.001 allowed)

----------------------------------------------------------------------
Ran 13 tests in 1.470s

FAILED (failures=8, skipped=1)
```
  
</details>

**Comment by Lezcano** [comment]
@IvanYashchuk any idea what could be causing this at all? I see that most of those tests fail just for a little bit, but there's a couple of them for which the error is still quite large...

**Comment by IvanYashchuk** [comment]
That test file is new, maybe the problem is somewhere in the implementation of this nested tensor mha?

**Comment by Lezcano** [comment]
cc @cpuhrsch as he might know about this.

I don't know much about `NestedTensor`s TBH. The only thing that occurs to me is that it may be dispatching (somehow) to some in-house implementation for `float16`. Either that, or `NestedTensor` may not be doing a good-enough job at accumulating some intermediate quantities?

**Comment by ezyang** [comment]
Is there anything I can do to help move this PR along?

**Comment by Lezcano** [comment]
ping @cpuhrsch about the failing nestedtensor tests (see https://github.com/pytorch/pytorch/pull/76828#issuecomment-1235432365).

**Comment by cpuhrsch** [comment]
@lezcano - Can you rebase this? The issue might already be resolved.

**Comment by cpuhrsch** [comment]
The failure in question might actually be legitimate signal that this change is causing a bug? One of the failures is

```
______________________________________ TestMHADeviceTypeCUDA.test_native_multihead_self_attention_cuda_float16 ______________________________________
Traceback (most recent call last):
  File "/scratch/cpuhrsch/pytorch/test/test_native_mha.py", line 274, in test_native_multihead_self_attention
    self._test_multihead_attention_impl(
  File "/scratch/cpuhrsch/pytorch/test/test_native_mha.py", line 247, in _test_multihead_attention_impl
    torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
  File "/scratch/cpuhrsch/pytorch/torch/testing/_comparison.py", line 1342, in assert_close
    assert_equal(
  File "/scratch/cpuhrsch/pytorch/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 9 / 8192 (0.1%)
Greatest absolute difference: 0.00390625 at index (9, 1, 58) (up to 0.001 allowed)
Greatest relative difference: 0.013805522208883553 at index (9, 1, 56) (up to 0.001 allowed)
```

The function in question (`_native_multi_head_attention`) uses matmul. @lezcano - what do we have that this is due to how matmul is used within this function as opposed to how matmul is changed in this PR? Maybe our other matmul unit tests don't provide enough signal. I'm happy to chat separately about how exactly matmul is being used in it. The test also fails when you disable the usage of NestedTensor, so I don't think it's related to the data structure.

**Comment by lezcano** [comment]
I agree, I think this is an internal cuBLAS issue, and that was my working hypothesis. Now, I wanted to check with you in case you could provide some more insight.
I think Ivan and I can take it from here. Thank you for looking into this though!

**Comment by lezcano** [comment]
OMG THE TEST FAILURES ARE GONE! The failures are unrelated. If this isn't a Christmas gift, I don't know what is.

@ngimel can we merge this one?

**Comment by lezcano** [comment]
when this one's merged, I'll put up a PR reverting https://github.com/pytorch/pytorch/pull/82662 and updating the decomposition of matmul accordingly.

**Comment by ngimel** [comment]
Ok let's merge if test failures are gone.

**Comment by lezcano** [comment]
I'll wait a few days to put up the other PRs in case this one is reverted.

**Comment by ngimel** [comment]
I don't think though we can revert #82662, the perf regression is real and it's not related to copies in matmul (only part of it is related at best).

**Comment by lezcano** [comment]
@pytorchbot merge -f "tests were already run"

it looks like it wasn't this PR that was causing that failure. It looks like it's a flaky test.

**Comment by mehtanirav** [comment]
This is causing some internal tests to fail with the following error.

```
terminate called after throwing an instance of 'c10::Error'
  what():  Could not find schema for aten::expand.
  
  
Exception raised from findSchemaOrThrow at xplat/caffe2/aten/src/ATen/core/dispatch/Dispatcher.cpp:131 (most recent call first):
(no backtrace available)

exit_code:"Process terminated by the signal: 1"
```

**Comment by mehtanirav** [comment]
Also causing assertion error in `FAIL: test_transformer_encode (executorch.test.end2end.test_end2end.DynamicShapeMemoryPlanningE2ETest)`

**Comment by mehtanirav** [comment]
@kit1980  Another PR related to LinAlg causing internal breakages.

**Comment by xwang233** [comment]
FYI, we're also seeing this PR at 0c3659586d (before reverted) breaking our internal benchmarks by using seemingly more GPU memories.

cc @ptrblck @ngimel

**Comment by lezcano** [comment]
@xwang233 this is something that we had already observed in https://github.com/pytorch/functorch/issues/989#issuecomment-1199807110

Is this something that you or anyone from nvidia could investigate? This PR *reduces* the number of copies that `matmul` performs in some cases (see the OP for which exact cases) so the fact that these changes consume more memory is very much unexpected.

**Comment by ngimel** [comment]
This PR changes some `bmm` calls into `mm` calls, and, due to functorch handling those ops differently they may end up on a more memory-consuming path during transformations.

**Comment by lezcano** [comment]
Do you (or @zou3519) have more details about that? It feels that's something we should be able to fix. 

Also, I'm interested in the internal NVIDIA test @xwang233 mentioned. Does that one also exercise functorch?

**Comment by xwang233** [comment]
It's a multi-GPU transformer training benchmark mostly based on https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/Translation/Transformer. The benchmark is in plain eager mode without jit or any other optimizations.

**Comment by zou3519** [comment]
Can we please not merge this until we verify that this doesn't lead to an OOM with functorch as well? The [script that OOMs](https://github.com/pytorch/functorch/issues/989#issuecomment-1199807110) is an important building block for many models that use functorch.

**Comment by zou3519** [comment]
> Do you (or @zou3519) have more details about that? It feels that's something we should be able to fix.

I'm happy to investigate further; I don't want to get into a situation where we do the PyTorch release and there is a large regression in functorch like the previous issue.

**Comment by ezyang** [comment]
ngl, if it is this important, we should have a unit test for it

**Comment by zou3519** [comment]
> ngl, if it is this important, we should have a unit test for it

We have benchmark tests in torchbench that look at performance numbers, but those are not entirely automated. It's unclear how to write a unittest here in the pytorch test suite, though we're taking suggestions

**Comment by ezyang** [comment]
In this particular case, I would propose having an expect test that traces through vmap over the relevant bmm call. This would be sufficient to detect if copies are happening.

**Comment by zou3519** [comment]
Some quick TL;DR on where we are with this. Longer doc over at https://docs.google.com/document/d/1_8TpPQyIgSpEbBID33jrkDyjQfN89ZPN8IKxwmO9kHQ/edit# .

- This PR is likely OK from a functorch perspective. I benchmarked it (see doc) on a P100 GPU and performance seems the same before/after. We should benchmark on an A100 GPU again to be safe (I didn't have time to grab an A100 GPU).
- The transformer problem reported by Nvidia above is a separate problem that needs investigation.
- After we land this PR, we cannot just land https://github.com/pytorch/pytorch/pull/75195 -- that one in conjunction with this PR causes 2+ problems. Hessian computation without functorch OOMs on a P100 GPU. Shrinking the model size to avoid OOM shows a significant regression in functorch performance (we should be able to write smarter batching rules for this, but it is a bit difficult because of how complicated the `matmul` implementation is)

**Comment by ngimel** [comment]
@zou3519 but that doesn't address concerns in https://github.com/pytorch/pytorch/pull/76828#issuecomment-1370374981, so probably this PR alone still increases mem usage in some situations and we shouldn't land it?
cc @xwang233

**Comment by zou3519** [comment]
> @zou3519 but that doesn't address concerns in [#76828 (comment)](https://github.com/pytorch/pytorch/pull/76828#issuecomment-1370374981), so probably this PR alone still increases mem usage in some situations and we shouldn't land it? cc @xwang233

That's correct, we still need to those investigate the memory issues

**Comment by lezcano** [comment]
FWIW, I'll follow up on Richard's work during his PTO

**Comment by xwang233** [comment]
I can run an internal CI test on this PR once it's in ready-to-go form. Benchmark may take a day or less. Feel free to ping me on this. Thanks!

**Comment by ngimel** [comment]
There are no changes to this PR since the last time merge was attempted, so if it was OOM'ing on transformers then, it will now.

**Comment by lezcano** [comment]
I'm able to repro the OOMs locally with a small repro, so I hope that if I fix that one I'll fix those from the internal tests. I'll ping you when this is ready for a second look though.

**Comment by lezcano** [comment]
Consider the repro from https://github.com/pytorch/functorch/issues/989#issue-1320568440
Patching in this PR and https://github.com/pytorch/pytorch/pull/75195/ I get an OOM in a GTX2060.
Now, I think this is not an issue coming from these PRs, but it's a latent issue from functorch that happened to be sidestepped by the bug that this PR is fixing.
Printing the shapes and strides of the matrices that hit `matmul` in that script, we see that we get a multiplication of two matrices o shapes `[512, 512] x [2, 512, 3]`. the first of them is transposed, but the second one is an insane slice with strides `[15360000, 30000, 1]`.
This matrix, with the optimisation from this PR, rather than being copied (and I guess that that'd free the massive matrix) is passed to `bmm`, which does not copy, and just performs the multiplication.

I've also observed that with the following change in the batching rule for `mv`, I get an OOM even earlier. I don't know if this is relevant, but leaving it here just in case it helps figuring out what pointers are being kept alive by whom
```diff
diff --git a/aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp b/aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp
index f26a4f79b14..9865e45396c 100644
--- a/aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp
+++ b/aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp
@@ -54,9 +54,9 @@ static std::tuple<Tensor, optional<int64_t>> tv_batch_rule(
   }
   else if (!self_bdim && other_bdim) {
     // ...OI, BI -> ...OI, IB -> OB
-    auto other_ = at::movedim(other, *other_bdim, -1);
-    auto result = at::matmul(self, other_);
-    return std::make_tuple( std::move(result), 1 );
+    auto other_ = moveBatchDimToFront(other, other_bdim);
+    auto result = at::matmul(self, other_.unsqueeze(-1)).squeeze(-1);
+    return std::make_tuple( std::move(result), 0 );
   }
   TORCH_INTERNAL_ASSERT(false, "can't get here");
 }
```

ccing the functorch gang @Chillee @kshitij12345 in case they can figure out the heck is going on here

**Comment by lezcano** [comment]
@xwang233 after further investigating this, I am tempted to think that the OOM issue may come from the fact that that given model was wrongly relying on the copy matmul was doing internally. My guess is that:
- There was some large matrix at some point in the model. 
- Then, just a slice of it was used and this was passed to `matmul`.
- Before, this slice was copied and the large matrix was deallocated. 
- Now, we don't copy that matrix but take a view, so the large matrix is not deallocated, and it OOMs

Could you have a look at that test see if this is in fact the case? If so, you should probably make a copy of the slice yourself, as this copying behaviour of `matmul` is not (and should not) be guaranteed.

**Comment by lezcano** [comment]
ping @xwang233

**Comment by xwang233** [comment]
Hi @lezcano , thanks for the update. I checked our transformer training script. There are indeed matrix multiplications of "a slice of a large tensor", but it's either a 2-D matrix slice passed to `torch.mm`, or a 3-D matrix slice passed to `torch.bmm`. There are also `view`s used in the script. For example, https://github.com/NVIDIA/DeepLearningExamples/blob/6a160116ef23c2f1c9497d50abc063eb01f0b9d9/PyTorch/Translation/Transformer/fairseq/modules/beamable_mm.py#L40-L46 and https://github.com/NVIDIA/DeepLearningExamples/blob/6a160116ef23c2f1c9497d50abc063eb01f0b9d9/PyTorch/Translation/Transformer/fairseq/modules/multihead_attention.py#L63-L94

Only `torch.mm` and `torch.bmm` are used in the training script and there is no usage of `torch.matmul`.

Does this match your assumptions in https://github.com/pytorch/pytorch/pull/76828#issuecomment-1412321258?

**Comment by lezcano** [comment]
That seems unrelated then. Note than this PR just touches matmul, not any other function.

**Comment by ngimel** [comment]
I don't understand, was then the conclusion that this PR broke transformer models with OOM incorrect?

**Comment by lezcano** [comment]
If https://github.com/pytorch/pytorch/pull/76828#issuecomment-1423344438 correct (they don't use matmul), then it should be. Actually, @xwang233, could you please run your internal CI on this PR see what's going on there?
And again, this PR just touches the `matmul` function, it does nothing to `bmm` and `mm`, so if the internal model does not use matmul (or the `@` operator), it cannot be that it affects your internal model...

**Comment by xwang233** [comment]
There is no matmul with the '@' operator in the code base.

I'll try to re-rerun our CI with your PR. Expecting results in a few days. Thanks for the update!

**Comment by xwang233** [comment]
Hi @lezcano , I was rebuilding your PR and testing on a transformer model, but I still _see OOM_ on a **8x** V100 16 GB machine. The training script uses `F.linear`, which uses `matmul` internally.  E.g. https://github.com/pytorch/pytorch/blob/0d7913c9c1217cf6d2770862f261d1af913fe63a/aten/src/ATen/native/Linear.cpp#L67

https://github.com/NVIDIA/DeepLearningExamples/blob/afea561ecff80b82f17316a0290f6f34c486c9a5/PyTorch/Translation/Transformer/fairseq/models/transformer.py#L333

https://github.com/NVIDIA/DeepLearningExamples/blob/afea561ecff80b82f17316a0290f6f34c486c9a5/PyTorch/Translation/Transformer/fairseq/modules/multihead_attention.py#L347-L348

I've tested your PR 
- [OOM] rebased at https://github.com/xwang233/pytorch/tree/76828-rebase @ https://github.com/xwang233/pytorch/commit/f2e16a565d26125aba0e8c6bf7c5b67f452016b4
- [not OOM] merge base with upstream master at f2aee8b8d526a04747a9c2f7b68e2caca611e5e8

Repro steps:
```bash
# build pytorch
git clone --recursive https://github.com/NVIDIA/DeepLearningExamples.git
cd DeepLearningExamples/PyTorch/Translation/Transformer/
git checkout afea561ecff80b82f17316a0290f6f34c486c9a5
git clone https://github.com/NVIDIA/cutlass.git && cd cutlass && git checkout ed2ed4d6 && cd ..
pip install -e .

python -m torch.distributed.run --nproc_per_node 8 train.py /data/wmt14_en_de_joined_dict_21501/ \
  --arch transformer_wmt_en_de_big_t2t \
  --share-all-embeddings \
  --optimizer adam \
  --adam-betas 0.9 0.997 \
  --adam-eps 1e-9 \
  --clip-norm 0.0 \
  --lr-scheduler inverse_sqrt \
  --warmup-init-lr 0.0 \
  --warmup-updates 4000 \
  --lr 0.0006 \
  --min-lr 0.0 \
  --dropout 0.1 \
  --weight-decay 0.0 \
  --criterion label_smoothed_cross_entropy \
  --label-smoothing 0.1 \
  --max-tokens 5120 \
  --seed 1 \
  --max-update 200 \
  --amp \
  --no-save \
  --save-dir results \
  --stat-file run_log_fp16.json
```

Tested with cuda 12.0 update 1, nvcc 12.0.140. On 8x V100 16 GB, at pytorch master branch, the training throughput is approximately 200,000 token/s.

The **1 GPU** version, by removing `-m torch.distributed.run --nproc_per_node 8` and using `python train.py ...` directly, also got OOM on the PR branch and trained successfully on pytorch master. Example error message
```python
Traceback (most recent call last):
  File "/workspace/DeepLearningExamples/PyTorch/Translation/Transformer/fairseq/ddp_trainer.py", line 217, in _backward
    self.scaler.scale(loss).backward()
  File "/opt/pytorch/pytorch/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "/opt/pytorch/pytorch/torch/autograd/__init__.py", line 200, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 14.92 GiB (GPU 0; 15.78 GiB total capacity; 4.68 GiB already allocated; 10.61 GiB free; 4.79 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by lezcano** [comment]
Yeah, that's annoying and it's happening again because those models may be passing slices of a larger model to `matmul`. I guess this is going to happen in other models in the wild. Again, I think this is caused by a subtle bug in these models. These models should **not** rely on `matmul` copying. In fact, most of the times it doesn't copy. This should be fixable by adding an explicit copy that would free the memory of the given matrix rather than being kept alive for the backwards pass.

As such, I don't think there is a way forward for this PR, as landing it would be too BC-breaking, even if we **should* be allowed to make that change. WDYT @ngimel?

**Comment by xwang233** [comment]
Does torch have a function that tells if a tensor is a slice of another tensor? If we add this temporarily to the PR branch and rebuild, we should be able to verify this assumption.

**Comment by lezcano** [comment]
I'm able to repro and I'm investigating the heck is going on here.
Interestingly enough, if you run the `train` step in the script under `with torch._dispatch.python.enable_python_dispatcher()` (which I was doing to enable debugging of some sort) the script doesn't OOM lol

**Comment by lezcano** [comment]
@ngimel @ezyang could you help landing this one?

**Pull request closed without merging**
